File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Collections/IQueue.cs
using System.Collections.Generic;

namespace BigEgg.Algorithm.Collections
{
    public interface IQueue<T> : IEnumerable<T>
    {
        /// <summary>
        /// Is this queue empty?
        /// </summary>
        /// <returns><c>True</c> if this queue is empty; <c>false</c> otherwise</returns>
        bool IsEmpty();

        /// <summary>
        /// Returns the number of items in this queue.
        /// </summary>
        /// <returns>The number of items in this queue.</returns>
        int Size();

        /// <summary>
        /// Adds the item to this queue.
        /// </summary>
        /// <param name="item">The item to add.</param>
        /// <exception cref="System.ArgumentNullException">Item cannot be null.</exception>
        void Enqueue(T item);

        /// <summary>
        /// Removes and returns the item on this queue that was least recently added.
        /// </summary>
        /// <returns>The item on this queue that was least recently added.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        T Dequeue();

        /// <summary>
        /// Returns (but does not remove) the item least recently added to this queue.
        /// </summary>
        /// <returns>The item least recently added to this queue.</returns>
        /// <exception cref="System.ArgumentOutOfRangeException">Queue is empty.</exception>
        T Peek();
    }
}

File: java/Algorithms/src/Collections/IQueue.java
package Collections;

import java.util.Iterator;

/**
 * Created by jianming.xiao on 10/7/14.
 */
public interface IQueue<Item> extends Iterable<Item> {
    /**
     * Is this queue empty?
     *
     * @return true if this queue is empty; false otherwise
     */
    boolean isEmpty();

    /**
     * Returns the number of items in this queue.
     *
     * @return the number of items in this queue
     */
    int size();

    /**
     * Adds the item to this queue.
     *
     * @param item the item to add
     * @throws java.lang.NullPointerException if the <tt>item</tt> is null;
     */
    void enqueue(Item item);

    /**
     * Removes and returns the item on this queue that was least recently added.
     *
     * @return the item on this queue that was least recently added
     * @throws java.util.NoSuchElementException if this queue is empty
     */
    Item dequeue();

    /**
     * Returns the item least recently added to this queue.
     *
     * @return the item least recently added to this queue
     * @throws java.util.NoSuchElementException if this queue is empty
     */
    Item peek();

    /**
     * Returns a string representation of this queue.
     *
     * @return the sequence of items in FIFO order, separated by spaces
     */
    String toString();

    /**
     * Returns an iterator that iterates over the items in this queue in FIFO order.
     *
     * @return an iterator that iterates over the items in this queue in FIFO order
     */
    Iterator<Item> iterator();
}
